### PR TITLE
[BEAM-1925] validate DoFn at pipeline creation time

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

- Move DoFnSignature from runner to transforms
- Add basic validation for process and run it inside Pardo
- No need to pass defaults to start / finish bundle

R: @chamikaramj PTAL

### CODE REVIEW

File: sdks/python/apache_beam/transforms/dofn_signature.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  Among other things, this will give an extensible way for for (1) accessing the
+  structure of the ``DoFn`` including methods and method parameters
+  (2) identifying features that a given ``DoFn`` support, for example, whether
+  a given ``DoFn`` is a Splittable ``DoFn`` (
+  https://s.apache.org/splittable-do-fn) (3) validating a ``DoFn`` based on the
+  feature set offered by it.
+  """
+
+  def __init__(self, do_fn):
+    # We add a property here for all methods defined by Beam DoFn features.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/dofn_signature.py]: Nit: Please move this log to the relevant place (3 lines below).

Comment (robertwb) [inline sdks/python/apache_beam/transforms/dofn_signature.py]: Or simply remove.

Comment (sb2nov) [inline sdks/python/apache_beam/transforms/dofn_signature.py]: Done

**Comment by sb2nov** [comment]
I'm seeing a change from 0.2 seconds on master vs 0.25 on this branch for the gist. This seems really odd to me as the only thing we're changing is the runner import.

@robertwb any insights?

**Comment by sb2nov** [comment]
Retest this please

**Comment by robertwb** [comment]
Can you make this change without moving stuff to a new file? Then it'll be clearer to see what's going on, and if the perf is OK we'll investigate why moving stuff around (and nothing else) is changing things.

**Comment by sb2nov** [comment]
Closing as #2963 caters to this too.

**Pull request closed without merging**